### PR TITLE
Made changes to links.json schema. Fixes #1285

### LINKED ISSUE #1275: Remove `input_bundles` from `analysis_process`
As specified in the [DCP2 MVP Schema doc](https://docs.google.com/document/d/1NsibP8g-NeLnksxlcBWQsSj5Zg_uimCDaAF1qB_qkjg/edit#heading=h.eeyjdjt197l9)

**For which schema is a change/update being suggested?**
`type/process/analysis_process.json`

**What should the change/update be?**

remove the field `input_bundles` ~and replace it with `input_files`~

* ~Field name: input_files~
* ~Field description: A list of references to the metadata entities that describe the individual input data files~
* ~Field type: array~
* ~item type: string~
* ~Required: yes~
* ~user_friendly: Input files~
* ~Examples: b816d2d6-5f10-4447-4194-3d0a804454d6~
* ~CV or enum: no~

### LINKED ISSUE #1285: modify links.json to allow for explicit supplementary file linking
**For which schema is a change/update being suggested?**

`system/links.json`

As part of work for DCP2 MVP we need a way of explicitly linking supplemental files to a project. The suggestion is to modify the links.json file to include both `experimental_graph_link` and `supplementary_file_link`.

Supplementary files should be able to be linked to any metadata entity whether that is the project, protocol or biomaterial.

**What should the change/update be?**

change the current `link` definition to `experimental_graph_link`, i.e. line 83-91 to read

```
"experimental_graph_link": {
      "type": "object",
      "required": [
        "process_type",
        "process_id",
        "inputs",
        "outputs",
        "protocols"
      ]
...
```

add a new type of `link` object, `supplementary_file_link`

```
"supplementary_file_link": {
      "type": "object",
      "required": [
        "inputs",
        "file_entity_id"
      ],
      "properties": {
        "inputs": {
          "description": "An array of inputs for this link",
          "type": "array",
          "items": {
            "type": "object",
            "$ref": "#/definitions/input"
          },
          "file_entity_id": {
            "description": "A link to a supplementary file uuid"
            "type": "string"
        }
```

modify the "links" definition to allow objects of both the new link types e.g.

```
"links": {
      "description": "An array of linking objects.",
      "type": "array",
      "items": {
        "type": "object",
        oneOf:
          - "$ref": "#/definitions/experimental_graph_link"
          - "$ref": "#/definitions/supplementary_file_link"
      }
```
tag @ESapenaVentura @rolando-ebi

### PR DESCRIPTION
**Pull request by ESapenaVentura**
### Release notes

For `system/links.json`:
- Added definitions and non-required `supplementary_file_link` property within `links` to allow for supplementary file linking.
 

### Reviews requested

- Need 2 Reviewers to approve because this is a major update
- This update will be applied as a hotfix once it's approved

### CODE REVIEW

**Comment by ESapenaVentura** [comment]
This change will be hotfixed for every environment as it's necessary for the DCP2 MVP

**Comment by ESapenaVentura** [comment]
@hannes-ucsc We've modified the PR to get both the requested change for the `analysis_process` schema and further changes to `links.json` schema to allow linking of supplementary files to any entity (Although for now all supplementary files will be linked to the project).

Please let us know if you'd like to suggest changes

File: json_schema/system/links.json [hunk 1]
```diff
@@ -80,7 +80,7 @@
         }
       }
     },
-    "link": {
+    "experiment_link": {
```
Comment (hannes-ucsc) [inline json_schema/system/links.json:83]: I would use

```suggestion
    "process_link": {
```

"experiment" may be overly restrictive. After all, each link represents a process.

File: json_schema/system/links.json [hunk 2]
```diff
@@ -128,33 +128,103 @@
...
+            "enum": [
+              "links"
+            ]
+          },
+          "links": {
+            "description": "An array of linking objects.",
+            "type": "array",
+            "items": {
+              "type": "object",
+              "oneOf": [
```
Comment (hannes-ucsc) [inline json_schema/system/links.json:219]: Since you are introducing polymorphism, you should consider adding a discriminator property to make it easier for metadata consumers to distinguish the types. Without a discriminator, consumers would need to apply either a heuristic (probing for certain distinguishing properties) or a full schema validation in order to determine which subtype of link they are dealing with. This could be done by adding a `link_type` property in each of the subschemas and pin it to either `process_link` or `supplementary_link` via a singleton enum.

File: json_schema/system/links.json [hunk 3]
```diff
@@ -128,33 +128,103 @@
...
+        }
+      }
     },
-    "links": {
-      "description": "An array of linking objects.",
-      "type": "array",
-      "items": {
-        "type": "object",
-        "$ref": "#/definitions/link"
+    "supplementary_file_link": {
```
Comment (hannes-ucsc) [inline json_schema/system/links.json:173]: Could shorten this to

```suggestion
    "supplementary_link": {
```

File: json_schema/system/links.json [hunk 4]
```diff
@@ -128,33 +128,103 @@
...
+      "properties": {
+         "entity_id": {
+            "description": "ID of this entity",
+            "example": "111f222a-333b-444c-555d-eee63ddfccc7",
+             "type": "string",
+             "pattern": "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"
+         },
+        "entity_type": {
+          "description": "The concrete type of this entity",
+          "type": "string"
```
Comment (hannes-ucsc) [inline json_schema/system/links.json:169]: I think this should be an enum or a pattern and explicitly list the concrete types that we, the DCP, support at this time. This, by the way, applies to all of the `…_type` properties. You could also implement the restrictions in a separate graph validator but not every consumer has access to that. JSONSchema is a standard and has off-the-shelves implementations, so we should use it to express graph constraints whenever possible. Our transitioning to concrete types in the previous PR for `links.json` enables us to do so.

**Comment by hannes-ucsc** [comment]
A note on process, if I may. You may want to factor out the `input_file` changes into a separate PR. These two changes have little to do with each other. One of them is less intrusive than the other and risks being taken hostage by a the discussion on that other change.

**Comment by ESapenaVentura** [comment]
Based on the comments by @hannes-ucsc above, we are going to keep this PR with only the `links.json` schema changes for further discussion and move the changes to `analysis_process.json` schema changes to a new PR.

File: json_schema/system/links.json (hunk 3 above)
Comment (ESapenaVentura) [reply json_schema/system/links.json:173]: Wouldn't this suggest a link that is supplementary, and not a link to a supplementary file?

**Comment by rolando-ebi** [comment]
@hannes-ucsc 

> Generally this goes in the right direction. But for DCP/2 MVP, we need to think how we would rewrite the existing `links.json` in v1 to this new schema. The only thing we know that a supplementary file (SF) is associated with a bundle and indirectly to a project. 

Yeah this would involve having to try to parse this info out of a v1 bundle. All v1 bundles should have a project. For any bundle that contains supplementary files (many do not), we can assume that they were all supplementary files for the project since ingest only allows supplementary files for project entities. So each v1 bundle would contain, at most, only one `supplementary_file_link`:

```
"links": [
  ...
  ...
  {
    "entity": {
        "entity_type": "project"
        "entity_id": "<project uuid>"
    },
    "supplementary_files": [
       <each file-metadata in the bundle with schema_type `supplementary_file`>
    ]
  }
]

```

File: json_schema/system/links.json (hunk 3 above)
Comment (hannes-ucsc) [reply json_schema/system/links.json]: It would and I think that's ok. I can't imagine a scenario in which a link to a supplementary file is not supplementary itself. MW says "supplementary: available to supply something extra when needed" so I think supplementary-ness is transitive, so to speak.

However, this is a cosmetic issue for me. Either name works for me.

**Comment by hannes-ucsc** [comment]
@rolando-ebi I completely agree. @danielsotirhos would be implementing that aspect of rewriting links.json. If you accept my suggestion of restricting `links[*].entity_type` we could initially just permit `project` as a value for that property and open it up as needed.

File: json_schema/property_migrations.json [hunk 5]
```diff
@@ -1,5 +1,14 @@
 {
     "migrations": [
+        {
```
Comment (mshadbolt) [inline json_schema/property_migrations.json]: should this still be in this pull request?

File: json_schema/system/links.json (hunk 1 above)
Comment (mshadbolt) [reply json_schema/system/links.json]: process link seems ok to me

File: json_schema/system/links.json (hunk 3 above)
Comment (mshadbolt) [reply json_schema/system/links.json]: I'd also prefer to keep `_file_` as it could be confusing as we have the concept of 'supplementary links' to link to websites/github etc from the project

File: json_schema/property_migrations.json (hunk 5 above)
Comment (ESapenaVentura) [reply json_schema/property_migrations.json]: It says to me that this is outdated and looking at the files this shouldn't appear anymore

**Comment by rolando-ebi** [comment]
@hannes-ucsc 

> If you accept my suggestion of restricting `links[*].entity_type` we could initially just permit `project` as a value for that property and open it up as needed later.

Yep, I think the constraint on the allowable entity_type values is a good idea. The wranglers would have to maintain this enum as the set of concrete entities change. Should be fairly straightforward to write a unit test to ensure none are missed. For now, will make it so that `project` is the only allowable entity_type. Note as well that `supplementary_file` is similarly the only permitted "file_type" in supplementary file links

File: json_schema/system/links.json (hunk 3 above)
Comment (hannes-ucsc) [reply json_schema/system/links.json]: Agreed.

**Comment by ESapenaVentura** [comment]
@kbergin Pinging you as this PR might be of interest to the analysis_team. Who should review this from your team?

Also PR #1286 might be of interest for your team to review as well

**Comment by ESapenaVentura** [comment]
@hannes-ucsc Rolando pushed the changes that were requested, does this PR sound reasonable?

File: json_schema/system/links.json [hunk 6]
```diff
@@ -1,160 +1,249 @@
...
+                "supplementary_files",
+                "link_type"
+            ],
+            "properties": {
+                "entity": {
+                    "description": "An entity that refers to this supplementary file",
+                    "type": "object",
+                    "$ref": "#/definitions/entity"
+                },
+                "supplementary_files": {
```
Comment (hannes-ucsc) [inline json_schema/system/links.json:197]:
```suggestion
                "files": {
```

Just a suggestion, to reduce the verbosity. Within a supplementary file link it should be clear that a linked file is a supplementary file.

Comment (ESapenaVentura) [reply json_schema/system/links.json:197]: LGTM!
